### PR TITLE
Shell: Allow * and ? wildcard expansion in arguments

### LINKED ISSUE #112: Shell: Support globbing (* expansion)
The shell should know how to expand asterisks (*).
If you're in a directory that contains the files "foo" and "bar", the shell should automagically expand "echo *" to "echo foo bar"

### PR DESCRIPTION
**Pull request by rburchell**
Should also presumably allow for escaping and such, but this is a start.
Fixes #112.

### CODE REVIEW

File: AK/String.cpp [hunk 1]
```diff
@@ -196,4 +196,48 @@ String String::repeated(char ch, int count)
     return *impl;
 }
 
+// TODO: case insensitive matching. consider adding a toupper/tolower implementation in String?
```
Comment (awesomekling) [inline AK/String.cpp]: String has to_uppercase() and to_lowercase(), is that what you had in mind?

File: AK/String.cpp [hunk 2]
```diff
@@ -196,4 +196,48 @@ String String::repeated(char ch, int count)
...
+bool String::matches(const String& mask) const
+{
+    if (is_null() || mask.is_null())
+        return false;
+
+    const char* string_ptr = (const char*)characters();
```
Comment (awesomekling) [inline AK/String.cpp]:
```suggestion
    const char* string_ptr = (const char*)characters();
````

File: AK/String.cpp [hunk 3]
```diff
@@ -196,4 +196,48 @@ String String::repeated(char ch, int count)
...
+    const char* mask_ptr = (const char*)mask.characters();
```
Comment (awesomekling) [inline AK/String.cpp]:
```suggestion
    const char* mask_ptr = mask.characters();
````

File: Shell/main.cpp [hunk 4]
```diff
@@ -5,13 +5,15 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include <dirent.h>
```
Comment (awesomekling) [inline Shell/main.cpp]: Do we actually need dirent.h?

File: Shell/main.cpp [hunk 5]
```diff
@@ -223,6 +225,48 @@ struct CommandTimer {
...
+                continue;
+            }
+
+            while (di.has_next()) {
+                String name = di.next_path();
+
+                // Dotfiles have to be explicitly requested
+                if (name[0] == '.' && arg[0] != '.') {
+                    continue;
+                } else {
```
Comment (awesomekling) [inline Shell/main.cpp]: I wouldn't use `else` after continue. It's a free chance to not nest :)

File: AK/String.cpp (hunk 1 above)
Comment (rburchell) [reply AK/String.cpp]: Hmm. I could have sworn I didn't see that :)

I can throw it in as an enum then already.

File: Shell/main.cpp (hunk 4 above)
Comment (rburchell) [reply Shell/main.cpp]: Right, we don't now I use CDirIterator ;)

**Comment by rburchell** [comment]
Yeah, I was thinking about variables (and ~) when I moved it over to a function :)

File: Shell/main.cpp [hunk 6]
```diff
@@ -223,6 +224,47 @@ struct CommandTimer {
...
+
+            while (di.has_next()) {
+                String name = di.next_path();
+
+                // Dotfiles have to be explicitly requested
+                if (name[0] == '.' && arg[0] != '.')
+                    continue;
+
+                // And even if they are, skip . and ..
+                if (name == "."  || name == "..")
```
Comment (awesomekling) [inline Shell/main.cpp]: bash includes "." and ".." in "echo .*"
IIRC you were using zsh, are you getting a different result there?

File: AK/String.cpp [hunk 7]
```diff
@@ -196,4 +196,58 @@ String String::repeated(char ch, int count)
     return *impl;
 }
 
+bool String::matches(const String& mask, CaseSensitivity case_sensitivity) const
+{
+    if (case_sensitivity == CaseSensitivity::CaseInsensitive) {
+        String this_lower = this->to_lowercase();
+        String mask_lower = mask.to_lowercase();
+        return this_lower.match_helper(mask_lower);
```
Comment (awesomekling) [inline AK/String.cpp]: In the glorious future we could have a version of this that doesn't create a bunch of new strings just for the comparison.

File: Shell/main.cpp (hunk 6 above)
Comment (rburchell) [reply Shell/main.cpp]: Yes indeed. Filtered out here :)

I think that filtering them is a nice thing to do unless we absolutely have to include them. They are really rather useless.

File: AK/String.cpp (hunk 7 above)
Comment (rburchell) [reply AK/String.cpp]: Yes, agreed. match_helper used to operate on the lowercase version of the mask/string inline, I was just too lazy to do so now :)

**Comment by awesomekling** [comment]
Ok! :+1:

**Pull request merged**